doc/sdcv.1
.TH SDCV 1 "2006-04-24" "sdcv-0.4.2"
.SH NAME
sdcv \- console version of StarDict program
.SH SYNOPSIS
.B sdcv 
[ 
.BI options 
] 
[list of words]
.SH DESCRIPTION
.I sdcv 
is a simple, cross-platform text-based utility
for working with dictionaries in StarDict format.
Each word from "list of words" may be a string
with a leading '/' for using a Fuzzy search algorithm,
with a leading '|' for using full-text search,
and the string may contain '?' and '*' for regexp search.
It works in interactive and non-interactive mode.
To exit from interactive mode press Ctrl+D. 
In interactive mode, 
if sdcv was compiled with readline library support,
you can use the UP and DOWN keys to cycle through history.
.SH OPTIONS
.TP 8
.B "\-h  \-\-help"
Display help message and exit
.TP 8
.B "\-v \-\-version"
Display version and exit
.TP 8
.B "\-l \-\-list\-dicts" 
Display list of available dictionaries and exit
.TP 8
.B "\-u \-\-use\-dict filename"
For search use only dictionary with this bookname
.TP 8
.B "\-n \-\-non\-interactive"
For use in scripts
.TP 8
.B "\-x \-\-only\-data\-dir"
For use in scripts: only use the dictionaries in data-dir, do not search in user and system directories
.TP 8
.B "\-e \-\-exact\-search" 
Do not fuzzy-search for similar words, only return exact matches
.TP 8
.B "\-j \-\-json"
Print the results of list-dicts and searches as json, not as plain text.
For use in automatically processing the results of a dictionary lookup.
.TP 8
.B "\-\-utf8\-output"
Force sdcv to not convert to locale charset, output in utf8
.TP 8
.B "\-\-utf8\-input"
Force sdcv to not convert from locale charset, assume that
input is in utf8
.TP 8
.B "\-\-data\-dir path/to/directory" 
Use this directory as the path to the stardict data directory. This means that
sdcv searches for dictionaries in data-dir/dic directory.
.TP 8
.B "\-\-color" 
Use ANSI escape codes for colorizing sdcv output (does not work with json output).
.SH FILES
.TP 
/usr/share/stardict/dic
.TP
$(XDG_DATA_HOME)/stardict/dic

Place where sdcv expects to find dictionaries.
Instead of /usr/share/stardict/dic you can use any directory
you want, just set the STARDICT_DATA_DIR environment variable.
For example, if you have dictionaries in /mnt/data/stardict-dicts/dic,
set STARDICT_DATA_DIR to /mnt/data/stardict-dicts.
.TP
$(XDG_DATA_HOME)/sdcv_history

This file includes the last $(SDCV_HISTSIZE) words, which you sought with sdcv.
SDCV uses this file only if it was compiled with readline library support.
.TP
$(XDG_CONFIG_HOME)/sdcv_ordering

This is a text file containing one dictionary bookname per line.
It specifies in which order the results of a search should be shown.
.SH ENVIRONMENT 
Environment Variables Used By \fIsdcv\fR:
.TP 20
.B STARDICT_DATA_DIR
If set, sdcv uses this variable as the data directory, this means that sdcv
searches dictionaries in $\fBSTARDICT_DATA_DIR\fR/dic
.TP 20
.B SDCV_HISTSIZE
If set, sdcv writes in $(XDG_DATA_HOME)/sdcv_history (or $(SDCV_HISTFILE)) the last $(SDCV_HISTSIZE) words,
which you look up using sdcv. If it is not set, then the last 2000 words are saved in $(XDG_DATA_HOME)/sdcv_history.
.TP 20
.B SDCV_HISTFILE
If set, sdcv writes it's history to $(SDCV_HISTFILE). If it is not set, then the default $(XDG_DATA_HOME)/sdcv_history path will be used.
.TP 20
.B SDCV_PAGER
If SDCV_PAGER is set, its value is used as the name of the program
to use to display the dictionary article.
.SH BUGS
Email bug reports to dushistov at mail dot ru. Be sure to include the word
"sdcv" somewhere in the "Subject:" field.
.SH AUTHORS
Evgeniy A. Dushistov, Hu Zheng
.SH SEE ALSO
stardict(1), http://sdcv.sourceforge.net/, http://stardict.sourceforge.net 


NEWS
Version 0.5.2
Synonyms index support (.syn files) by Peter <craven@gmx.net>
Add support of json output by Peter <craven@gmx.net> (--json-output)
Add -e for exact searches (no fuzzy matches) by Peter <craven@gmx.net>
Fix build with clang 3.4.1
fix FSF address in license by Tomáš Čech <sleep_walker@suse.com>

Version 0.5.1
 Fix usage of SDCV_PAGER by Anton Yuzhaninov
 Fix build without readline<
 specify dictionary order by -u switches or ~/.sdcv_ordering by Cong Gu

Version 0.5
- add option to colorize output
- Roman Imankulov's patch to better use of readline
- Fix build with modern compilers
- Fix work on CPU with align issues, like ARM and SPARC (thanks to Michal Čihař )
- Migrate to C++11 and CMake (so at now archive with code 180K->54K)
- Update Russian translation

Version 0.4.3
- fix build problems with gcc 4.x
- French translation, thanks to Vincent Petry <PVince81@yahoo.fr>
- Slovak translation update, thanks to Ivan Masár <helix84@centrum.sk>
- Add Czech translation, thanks to Michal Čihař <michal@cihar.com>

Version 0.4.2
* Cache scheme has been rewritten, size of cache has been reduced
* Merge with stardict, speedup of lookup, and add full-text search
* Russian translation update

Version 0.4.1
* Recreate cache if idx file was modified 
* Abbility to use pager(SDCV_PAGER)
* Add Chinese (traditional)  translation
* Add  Ukrainian translation

Version 0.4
* Fix several bugs
* libstardict by default linked statically with sdcv
* Add a Simplified Chinese translation
* documentation update
* add ability to use readline library with sdcv
* add ability hold history in ~/.sdcv_history

Version 0.3.4
* Fix bug of handling of fuzzy search
* More clever showing search results

Version 0.3.3
* Fix bug in using -u option
* Update Slovak translation

Version 0.3.2
* Fix bug in Dictionary::load for correct work with gzip'ed idx files
* Add Slovak translation
* Change format of output for -l option

Version 0.3.1
* Fix bug in Lib::Lookup, which cause SF and other problems in search
* Change installation procedure to install with libstardict headers files,
* to help compile any application with libstardict
* Documentation update

Version 0.3
* update libstardict, make search in dictionaries and loading of dictionaries
faster
* add --data-dir option
* -u, --use-dict now take bookname as argument, not a name of 'ifo' file
* documentation update and improve

Version 0.2

* Add autoconf, automake, gettext support.
* Divide package into two parts: libstardict and sdcv by it self.
* Add options for:
*   search word in special dictionary
*   get list of dictionaries
*   for using in scripts


src/sdcv.cpp
/*
 * This file part of sdcv - console version of Stardict program
 * http://sdcv.sourceforge.net
 * Copyright (C) 2003-2006 Evgeniy <dushistov@mail.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <algorithm>
#include <cerrno>
#include <clocale>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <map>
#include <memory>
#include <string>
#include <vector>

#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gstdio.h>

#include "libwrapper.hpp"
#include "readline.hpp"
#include "utils.hpp"

static const char gVersion[] = VERSION;

namespace
{
static void free_str_array(gchar **arr)
{
    gchar **p;

    for (p = arr; *p; ++p)
        g_free(*p);
    g_free(arr);
}
}
namespace glib
{
using StrArr = ResourceWrapper<gchar *, gchar *, free_str_array>;
}

static void list_dicts(const std::list<std::string> &dicts_dir_list, bool use_json);

int main(int argc, char *argv[]) try {
    setlocale(LC_ALL, "");
#if ENABLE_NLS
    bindtextdomain("sdcv",
                   //"./locale"//< for testing
                   GETTEXT_TRANSLATIONS_PATH //< should be
                   );
    textdomain("sdcv");
#endif

    gboolean show_version = FALSE;
    gboolean show_list_dicts = FALSE;
    glib::StrArr use_dict_list;
    gboolean non_interactive = FALSE;
    gboolean json_output = FALSE;
    gboolean no_fuzzy = FALSE;
    gboolean utf8_output = FALSE;
    gboolean utf8_input = FALSE;
    glib::CharStr opt_data_dir;
    gboolean only_data_dir = FALSE;
    gboolean colorize = FALSE;

    const GOptionEntry entries[] = {
        { "version", 'v', 0, G_OPTION_ARG_NONE, &show_version,
          _("display version information and exit"), nullptr },
        { "list-dicts", 'l', 0, G_OPTION_ARG_NONE, &show_list_dicts,
          _("display list of available dictionaries and exit"), nullptr },
        { "use-dict", 'u', 0, G_OPTION_ARG_STRING_ARRAY, get_addr(use_dict_list),
          _("for search use only dictionary with this bookname"),
          _("bookname") },
        { "non-interactive", 'n', 0, G_OPTION_ARG_NONE, &non_interactive,
          _("for use in scripts"), nullptr },
        { "json-output", 'j', 0, G_OPTION_ARG_NONE, &json_output,
          _("print the result formatted as JSON"), nullptr },
        { "exact-search", 'e', 0, G_OPTION_ARG_NONE, &no_fuzzy,
          _("do not fuzzy-search for similar words, only return exact matches"), nullptr },
        { "utf8-output", '0', 0, G_OPTION_ARG_NONE, &utf8_output,
          _("output must be in utf8"), nullptr },
        { "utf8-input", '1', 0, G_OPTION_ARG_NONE, &utf8_input,
          _("input of sdcv in utf8"), nullptr },
        { "data-dir", '2', 0, G_OPTION_ARG_STRING, get_addr(opt_data_dir),
          _("use this directory as path to stardict data directory"),
          _("path/to/dir") },
        { "only-data-dir", 'x', 0, G_OPTION_ARG_NONE, &only_data_dir,
          _("only use the dictionaries in data-dir, do not search in user and system directories"), nullptr },
        { "color", 'c', 0, G_OPTION_ARG_NONE, &colorize,
          _("colorize the output"), nullptr },
        {},
    };

    glib::Error error;
    GOptionContext *context = g_option_context_new(_(" words"));
    g_option_context_set_help_enabled(context, TRUE);
    g_option_context_add_main_entries(context, entries, nullptr);
    const gboolean parse_res = g_option_context_parse(context, &argc, &argv, get_addr(error));
    g_option_context_free(context);
    if (!parse_res) {
        fprintf(stderr, _("Invalid command line arguments: %s\n"),
                error->message);
        return EXIT_FAILURE;
    }

    if (show_version) {
        printf(_("Console version of Stardict, version %s\n"), gVersion);
        return EXIT_SUCCESS;
    }

    const gchar *stardict_data_dir = g_getenv("STARDICT_DATA_DIR");
    std::string data_dir;
    if (!opt_data_dir) {
        if (!only_data_dir) {
            if (stardict_data_dir)
                data_dir = stardict_data_dir;
            else
                data_dir = "/usr/share/stardict/dic";
        }
    } else {
        data_dir = get_impl(opt_data_dir);
    }

    std::string conf_dir = std::string(g_get_home_dir()) + G_DIR_SEPARATOR + ".stardict";
    if (!g_file_test(conf_dir.c_str(), G_FILE_TEST_IS_DIR))
        conf_dir = std::string(g_get_user_data_dir()) + G_DIR_SEPARATOR + "stardict";

    std::list<std::string> dicts_dir_list;
    if (!only_data_dir)
        dicts_dir_list.push_back(conf_dir + G_DIR_SEPARATOR + "dic");
    dicts_dir_list.push_back(data_dir);
    if (show_list_dicts) {
        list_dicts(dicts_dir_list, json_output);
        return EXIT_SUCCESS;
    }

    std::list<std::string> disable_list;

    std::map<std::string, std::string> bookname_to_ifo;
    for_each_file(dicts_dir_list, ".ifo", std::list<std::string>(), std::list<std::string>(),
                  [&bookname_to_ifo](const std::string &fname, bool) {
                      DictInfo dict_info;
                      const bool load_ok = dict_info.load_from_ifo_file(fname, false);
                      if (!load_ok)
                          return;
                      bookname_to_ifo[dict_info.bookname] = dict_info.ifo_file_name;
                  });

    std::list<std::string> order_list;
    if (use_dict_list != nullptr) {
        for (auto &&x : bookname_to_ifo) {
            gchar **p = get_impl(use_dict_list);
            for (; *p != nullptr; ++p)
                if (x.first.compare(*p) == 0) {
                    break;
                }
            if (*p == nullptr) {
                disable_list.push_back(x.second);
            }
        }

        // add bookname to list
        gchar **p = get_impl(use_dict_list);
        while (*p) {
            order_list.push_back(bookname_to_ifo.at(*p));
            ++p;
        }
    } else {
        std::string ordering_cfg_file = std::string(g_get_user_config_dir()) + G_DIR_SEPARATOR_S "sdcv_ordering";
        FILE *ordering_file = fopen(ordering_cfg_file.c_str(), "r");
        if (ordering_file == nullptr) {
            ordering_cfg_file = std::string(g_get_home_dir()) + G_DIR_SEPARATOR_S ".sdcv_ordering";
            ordering_file = fopen(ordering_cfg_file.c_str(), "r");
        }
        if (ordering_file != nullptr) {
            std::string line;
            while (stdio_getline(ordering_file, line)) {
                order_list.push_back(bookname_to_ifo.at(line));
            }
            fclose(ordering_file);
        }
    }

    if (g_mkdir(conf_dir.c_str(), S_IRWXU) == -1 && errno != EEXIST) {
        fprintf(stderr, _("g_mkdir failed: %s\n"), strerror(errno));
    }

    Library lib(utf8_input, utf8_output, colorize, json_output, no_fuzzy);
    lib.load(dicts_dir_list, order_list, disable_list);

    std::unique_ptr<IReadLine> io(create_readline_object());
    if (optind < argc) {
		search_result rval = SEARCH_SUCCESS;
        for (int i = optind; i < argc; ++i)
            if ((rval = lib.process_phrase(argv[i], *io, non_interactive)) != SEARCH_SUCCESS) {
                return rval;
            }
    } else if (!non_interactive) {

        std::string phrase;
        while (io->read(_("Enter word or phrase: "), phrase)) {
            if (lib.process_phrase(phrase.c_str(), *io) == SEARCH_FAILURE)
                return EXIT_FAILURE;
            phrase.clear();
        }

        putchar('\n');
    } else {
        fprintf(stderr, _("There are no words/phrases to translate.\n"));
    }
    return EXIT_SUCCESS;
} catch (const std::exception &ex) {
    fprintf(stderr, "Internal error: %s\n", ex.what());
    exit(EXIT_FAILURE);
}

static void list_dicts(const std::list<std::string> &dicts_dir_list, bool use_json)
{
    bool first_entry = true;
    if (!use_json)
        printf(_("Dictionary's name   Word count\n"));
    else
        fputc('[', stdout);
    std::list<std::string> order_list, disable_list;
    for_each_file(dicts_dir_list, ".ifo", order_list,
                  disable_list, [use_json, &first_entry](const std::string &filename, bool) -> void {
                      DictInfo dict_info;
                      if (dict_info.load_from_ifo_file(filename, false)) {
                          const std::string bookname = utf8_to_locale_ign_err(dict_info.bookname);
                          if (use_json) {
                              if (first_entry) {
                                  first_entry = false;
                              } else {
                                  fputc(',', stdout); // comma between entries
                              }
                              printf("{\"name\": \"%s\", \"wordcount\": \"%d\"}", json_escape_string(bookname).c_str(), dict_info.wordcount);
                          } else {
                              printf("%s    %d\n", bookname.c_str(), dict_info.wordcount);
                          }
                      }
                  });
    if (use_json)
        fputs("]\n", stdout);
}
